### DISCUSSION TITLE
Parameter logging of multiple crazyflies is delayed

### DISCUSSION THREAD

**Discussion by SeppeVdB** [bitcraze/discussions#722] [Q&A]
I am working on implementing a swarm of crazyflies using the lighthouse external positioning system and the crazyradio2 for communication. I am currently working with 6 drones, each sending their positition, velocity, and yaw at 50Hz separated over two packets. So 100 packets per seconds per drone, is 600 packets per second in total. Reading other discussions, this is still under the limit of 800 packets.

The problem I'm running into is that the logging callbacks are only called at about 400Hz and that the logged data has a two seconds delay. It's not as if it is skipping logging parameters because, making a smooth motion with the drone, results in the same logged motion, only two seconds later. So no data is lost, only delayed.

If anyone would have an idea on what is causing this, I will be happy to hear it.

**Comment by krichardsson** [comment]
Which channels/addresses are you using?

**Reply by SeppeVdB** [reply]
All 6 drones are using channel 90 at 2M and have unique addresses

**Comment by SeppeVdB** [comment]
I have also noticed that the data rate/number of drones limit is different for different laptops. Maybe something that is useful to know is that I am running the code with ROS2 through WSL (Ubuntu22.04).

**Reply by ataffanel** [reply]
Hi,

Crazyradio 2 is still working in Crazyradio 1 USB-mode and this mode has a major limitation: the bandwidth is pretty much divided by 2 as soon as you are talking to more than one Crazyflie. This is something that will be fixed when we finish the new Crazyradio2 USB mode.

So what you are observing kind of make sense. I would suggest using more Crazyradios or reduce the number of Crazyflies to increase the available bandwidth.

As a side note, we have observed instabilities when using WSL, at least on Windows 10. So trying to use a VM or, even better, a native Linux install, could help with the performance consistency.

**Reply by SeppeVdB** [reply]
Thanks for the reply, this information is very useful.

We think the problem lies with the amount of threads we are using. Multiple threads are created to control the swarm and in combination with the threads running in the crazyflie library, causes a delay in runtime. We have tried using multiple radios but this doesn't solve the issue. If the threads are the problem, this result would make sense.

We are currently thinking of rewriting the radiodriver to reduce the amount of threads it's using. Cause we already have a separate thread per drone in our ROS2 software. So any tips related to this would be appreciated.

**Reply by ataffanel** [reply]
What language are you using, Python or C++? Thread handling in python can indeed be quite slow and cause problems. There should be less problems with C++.

I unfortunately do not have any tips other than trying with C++, Crazyswarm 2 is based on ROS2 and has a CPP back-end as far as I know.

I have been personally playing with making Rust-implemented lib that would solve scalability problem, but it is all still very experimental.

**Reply by SeppeVdB** [reply]
Ah yes we are using Python. I will have a look at crazyswarm2 and the rust implementation sounds interesting for the future.

Thanks for the help

**Comment by SeppeVdB** [comment]
After some testing I am getting the following data:

| num radios | num drones | packets/second/drone | packets/second/radio total |
| - | - | - | - |
| 1 | 1 | 100 | 100 |
| 1 | 2 | 100 | 200 |
| 1 | 3 | 100 | 300 |
| 1 | 4 | 100 | 400 |
| 1 | 5 | 50 | 250 |
| 1 | 6 | 50 | 300 |
| 1 | 7 | 50 | 350 |
| 1 | 8 | 33 | 264 |
| 1 | 9 | 33 | 297 |
| 1 | 10 | 25 | 250 |
| 2 | 2 | 100 | 100|
| 2 | 4 | 100 | 200|
| 2 | 6 | 100 | 300 |
| 2 | 8 | 50 | 200 |
| 2 | 10 | 50 | 250 |

These are the maximum data rates at which no delay in the data was found. It looks like the data rates are divisions of 100. Does this data make sense to you? And is this the result of a software or hardware limit?

**Reply by krichardsson** [reply]
@ataffanel is at ICRA and he will probably not have time to answer until next week

**Reply by ataffanel** [reply]
Hi, sorry for the late answer, we where at a conference and I am just getting back to date with what happens last week.

Thanks a lot for the measurements. 1 radio with 100 packets per seconds matches exactly my expectation: I would expect the limit to be at about 400 packets per seconds for big (ie. almost full) packets.

Your other measurement tends to show a problem with the software part, the more Crazyflie object the less bandwidth we get. The radio hardware does not know about nor care about the number of Crazyflies. Though, the USB protocol is really not helping for that either since it forces Python to handle the full communication scheduling thus stressing it.

Is is possible to provide me the code you use for these tests? I would be very interested to run the same test with the Rust lib as well as with the new Crazyradio2 USB protocol. This would allow to verify where the limitation is.

**Reply by SeppeVdB** [reply]
I am currently quite busy, so I will try to give you a minimal working example for these tests somewhere next week. I'll also measure the limits for more than 10 drones and more than 2 radios.

The packets I was receiving during these tests are almost full packets: position and velocity, so 6 floats.

I am also currently using a usb hub to use multiple radios on one usb port. I haven't done extensive testing using this usb hub, but at first sight it is giving the same performance as using separate usb ports per radio. Maybe this information is also of use.

**Reply by ataffanel** [reply]
For the USB-hub there is a theoretical limit that I have never measured practically: Crazyradio is using 'Full Speed USB 2' at 12Mbps (Sometime called USB 1.1 since this was the max speed of USB 1.1 as well). When High-speed USB (480Mbps) came with USB2 they used the same data line so hubs talk using High-speed to the PC and there is a 'transaction translator' (TT) that converts 480MBps to and from 12 or 1.5Mbps.

Some hub share one TT for all port which means that you will only get 12Mbps shared over all ports. Some hub have one TT per port though which is much better since you can basically get the full 12Mbps on all ports. In theory, one TT per port should have much better performance. I am not sure how much it affects Crazyradio operation though.

In Linux it is possible to see what architecture the hub has:
```
~
❯ sudo lsusb -v -d 0451:8442 | grep bDeviceProtocol | head -1      
  bDeviceProtocol         2 TT per port

~
❯ sudo lsusb -v -d 1d6b:0002 | grep bDeviceProtocol | head -1      
  bDeviceProtocol         1 Single TT
```

Separate port on a modern laptop will usually be driven by different USB bus so they will provide one TT per port.
